### ISSUE TITLE
undefined variable 'isPath' in pkgs/build-support/coq/meta-fetch/default.nix

### ISSUE THREAD

**Issue by telotortium** [NixOS/nixpkgs#116670]
**Describe the bug**

Evaluating any package on my installation (to install it, for example) throws an `error: undefined variable 'isPath' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/meta-fetch/default.nix:66:12`.

```
$ nix-env --show-trace -i hello
error: while evaluating the attribute 'coq' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/top-level/coq-packages.nix:124:3:
while evaluating 'recurseIntoAttrs' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/lib/attrsets.nix:502:5, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/top-level/coq-packages.nix:121:17:
while evaluating 'mapDerivationAttrset' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/lib/meta.nix:46:29, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/top-level/coq-packages.nix:121:35:
while evaluating 'mkCoqPackages' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/top-level/coq-packages.nix:98:19, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/top-level/coq-packages.nix:118:22:
while evaluating 'filterPackages' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/top-level/coq-packages.nix:72:24, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/top-level/coq-packages.nix:100:5:
while evaluating 'filterCoqPackages' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/top-level/coq-packages.nix:75:23, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/top-level/coq-packages.nix:72:55:
while evaluating anonymous function at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/top-level/coq-packages.nix:77:22, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/top-level/coq-packages.nix:77:7:
while evaluating 'optional' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/lib/lists.nix:254:20, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/top-level/coq-packages.nix:78:11:
while evaluating the attribute 'meta.coqFilter' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:372:19:
while evaluating the attribute 'meta' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/default.nix:75:3:
while evaluating 'optionalAttrs' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/lib/attrsets.nix:349:25, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/default.nix:80:5:
while evaluating anonymous function at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/meta-fetch/default.nix:36:1, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/default.nix:45:13:
while evaluating 'switch' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:139:26, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/meta-fetch/default.nix:37:1:
while evaluating 'switch-if' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:111:18, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:144:5:
while evaluating 'findFirst' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/lib/lists.nix:199:5, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:111:22:
while evaluating 'combine' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:141:21, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:144:34:
while evaluating 'opTruncate' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:7:28, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/development/coq-modules/Cheerios/default.nix:7:23:
while evaluating 'versionAtLeast' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/lib/strings.nix:503:24, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:8:10:
while evaluating 'versionOlder' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/lib/strings.nix:491:22, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/lib/strings.nix:503:29:
while evaluating 'truncate' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:6:21, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:8:14:
while evaluating 'sublist' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/lib/lists.nix:589:5, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:6:46:
while evaluating the attribute 'coq-version' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/applications/science/logic/coq/default.nix:76:12:
while evaluating anonymous function at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/meta-fetch/default.nix:36:1, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/applications/science/logic/coq/default.nix:48:13:
while evaluating 'switch' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:139:26, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/meta-fetch/default.nix:37:1:
while evaluating 'switch-if' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:111:18, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:144:5:
while evaluating 'findFirst' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/lib/lists.nix:199:5, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:111:22:
while evaluating 'combine' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:141:21, called from /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/extra-lib.nix:144:34:
undefined variable 'isPath' at /nix/store/2d7k89kyd69iakyrwhbrhkl7mrzg5c4h-nixpkgs-21.05pre276671.94a0cb3b8d8/nixpkgs/pkgs/build-support/coq/meta-fetch/default.nix:66:12
```

**Notify maintainers**
@siraben @CohenCyril - contributors to coq/meta-fetch/default.nix - in particular https://github.com/NixOS/nixpkgs/commit/82454e9df8d54c5afaa23055b4fbc656e347c71e looks suspicious.

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.7.17-1rodete5-amd64, Debian GNU/Linux, noversion`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.2.2`
 - channels(rmirelan): `"nixpkgs-21.05pre276671.94a0cb3b8d8"`
 - channels(root): `"nixpkgs-21.05pre276671.94a0cb3b8d8"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by CohenCyril** [comment]
I cannot reproduce... can you try #16674

**Comment by telotortium** [comment]
@CohenCyril Thanks for the quick fix! That seems to work - the following command succeeds (`$HOME/misc/build/nixpkgs` is where I've cloned your `fix-116670` branch:

```
$ nix-env -I nixpkgs=$HOME/misc/build/nixpkgs -i hello
installing 'hello-2.10'
```

**Comment by CohenCyril** [comment]
OK, I'll take further steps later on to get it merged.

BTW out of curiosity what is the result of 
```bash
nix eval '({x}: let x.y = true; in with x; y) {x.y = false;}'
```
with your version of nix?

**Comment by telotortium** [comment]
```
$ nix --version
nix (Nix) 2.2.2
$ nix eval '({x}: let x.y = true; in with x; y) {x.y = false;}'
true
```

**Comment by telotortium** [comment]
Should I have updated Nix to the latest version (2.3.10)? I was able to work around it by changing my nixpkgs channel to nixos-20.09 and then updating the `nix` package.

**Comment by CohenCyril** [comment]
Nixpkgs works with and without my fix on Nix 2.3.10, I don't understand what exactly is the problem with Nix 2.2.2
